### PR TITLE
Finish initial pass on wasm-binary, and add wasm-as tool #13

### PR DESCRIPTION
**Pull request by kripken**
General idea is `wasm-as` is a parallel to `llvm-as`, and so forth. But no strong feelings about names, and perhaps anyhow all tools should be able to load both binary and text eventually.

This just adds the binary writer, and checks that it compiles.

### CODE REVIEW

File: src/wasm-as.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2015 WebAssembly Community Group participants
```
Comment (jfbastien) [inline src/wasm-as.cpp]: 2016 now.

File: src/wasm-as.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                o->extra["output"] = argument;
+                Colors::disable();
+              })
+      .add_positional("INFILE", Options::Arguments::One,
+                      [](Options *o, const std::string &argument) {
+                        o->extra["infile"] = argument;
+                      });
+  options.parse(argc, argv);
+
+  std::string input(read_file(options.extra["infile"], options.debug));
```
Comment (jfbastien) [inline src/wasm-as.cpp]: I just changed the API to either return a `std::string` or `std::vector<char>`. Change to:
```
auto input(read_file<std::string>(options.extra["infile"], options.debug));
```

File: src/wasm-as.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  if (options.debug) std::cerr << "s-parsing...\n";
```
Comment (jfbastien) [inline src/wasm-as.cpp]: `std::endl` and `\n` mixture here and below is a bit odd. `std::endl` just flushes buffered output as well (`std::cerr` usually isn't so it's more of a C++ style thing than anything).

File: src/wasm-as.cpp (hunk 1 above)
Comment (kripken) [inline src/wasm-as.cpp]: fixed

File: src/wasm-as.cpp (hunk 2 above)
Comment (kripken) [inline src/wasm-as.cpp]: done

File: src/wasm-as.cpp (hunk 3 above)
Comment (kripken) [inline src/wasm-as.cpp]: fixed

**Comment by jfbastien** [comment]
Could you add the following to `check.py`:
```
import test.waterfall.src.link_assembly_files as link_assembly_files
wasm_as_torture_out = os.path.abspath(os.path.join('test', 'wasm-as-torture-out'))
if os.path.isdir(wasm_as_torture_out):
  shutil.rmtree(wasm_as_torture_out)
os.mkdir(wasm_as_torture_out)
unexpected_result_count = link_assembly_files.run(
    linker=os.path.abspath(os.path.join('bin', 'wasm-as')),
    files=os.path.abspath(os.path.join('test', 'torture-s', '*.s')),
    fails=os.path.abspath(os.path.join('test', 'wasm_as_known_gcc_test_failures.txt')),
    out=wasm_as_torture_out)
assert os.path.isdir(wasm_as_torture_out), 'Expected output directory %s' % wasm_as_torture_out
shutil.rmtree(wasm_as_torture_out)
if unexpected_result_count:
  fail(unexpected_result_count, 0)
```
It'll also require a new `test/wasm_as_known_gcc_test_failures.txt` file, and an update of the `test/waterfall` submodule.

**Comment by kripken** [comment]
Wait, what does that code do? Run wasm-as on stuff? It's not ready yet! :)

**Comment by jfbastien** [comment]
Not even mildly ready to fail on most-but-not-quite-all tests? :p

**Comment by kripken** [comment]
No, it aborts on "hello world" :)

I need to do some other refactoring. This pull is mainly to finish getting it building, as I don't want to had a long-term side branch.

**Comment by jfbastien** [comment]
Ah OK, lgtm then :)

**Comment by JSStats** [comment]
Just out of curiosity, the llvm backend currently emits a .s file in a format that is not vaguely like wast. This tools is describe as 'Assemble a .wast (WebAssembly text format) into a .wasm (WebAssembly binary format)' and if so then the wasm-as naming looks confusing. Is it planned that the current .s format becomes the text format and superseded wast or might there be a separate .s format?

File: src/wasm-as.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#include "support/command-line.h"
+#include "support/file.h"
+#include "wasm-binary.h"
+#include "wasm-s-parser.h"
+
+using namespace cashew;
+using namespace wasm;
+
+int main(int argc, const char *argv[]) {
+  Options options("wasm-as", "Assemble a .wast (WebAssembly text format) into a .wasm (WebAssembly binary format)");
```
Comment (jfbastien) [inline src/wasm-as.cpp]: "Link and assemble a .s into a .wasm"

**Comment by jfbastien** [comment]
> Just out of curiosity, the llvm backend currently emits a .s file in a format that is not vaguely like wast. This tools is describe as 'Assemble a .wast (WebAssembly text format) into a .wasm (WebAssembly binary format)' and if so then the wasm-as naming looks confusing.

It consumes a `.s` file, the help is wrong but I hadn't noticed. Added a comment.

> Is it planned that the current .s format becomes the text format and superseded wast or might there be a separate .s format?

The `.s` format is just convenient for LLVM as it is now, because that got things off the ground faster without introducing alien concepts into LLVM (which the LLVM community may frown upon). It's meant to look similar to `.s` files for other architectures, LLVM makes that easy. @sunfishcode is experimenting with outputting binary directly from LLVM, which is more involved but then you get assembler and disassembler for free if you Do It The Right Way™.

The plans for text format are still very up in the air, and the current s-expression format isn't agreed on either. `.s` seems less readable than `.wast` so I doubt we'll standardize it as The One True Text Format™, but it may be useful to standardize between compilers, the same way standardizing `.o` / `.a` / `.so` files between compilers could be useful: they're not meant to get shipped to a browser, but tools can get along and that's nice.

**Comment by kripken** [comment]
Actually no, the text as it stands is correct. This tool assembles a .wast (s-expression wasm text format, and eventually the official wasm text format whatever that will be) into a .wasm (binary wasm format, in current v8 proposal).

In other words:

 * `wasm-as: .wast => .wasm`
 * `wasm-dis: .wasm => .wast` - not yet implemented

These are parallel to `llvm-as, llvm-dis`. They translate between the text and binary formats of the same thing, either LLVM IR or wasm AST.

Other tools convert to/from external or unofficial formats, like

 * `asm2wasm: .asm.js = .wasm` (well, right now it emits text, but that should change)
 * `s2wasm: LLVM .s => .wasm` (again, right now it emits text)
 * etc.

A sad point of confusion is that `.s` is LLVM's .s format, while `wasm-s-parser` stands for `wasm-s-expression-parser`, which might trip people up. Suggestions for a better name?

**Comment by jfbastien** [comment]
@kripken ah right, I double-misread :-)

@binji was using `sexpr`, but in general `wasm-as` seems intuitive. So would `wast2wasm` and similar.

**Comment by binji** [comment]
@jfbastien I named it that before .wast existed :)

**Pull request merged**